### PR TITLE
Add auto peering for Consul Federation

### PR DESCRIPTION
**Pull request by i0rek**
### :hammer_and_wrench: Description

This PR adds the new parameter `auto_hvn_to_hvn_peering` to consul clusters. It enables the user to decided if they want to use automatic peering during federation or not. Needs https://github.com/hashicorp/cloud-consul-service/pull/1253.

This PR also semi depends on https://github.com/hashicorp/terraform-provider-hcp/pull/156 because right now there is no way to create a manual peering.

### :building_construction: Acceptance tests

- [ ] Are there any feature flags that are required to use this functionality?
- [ ] Have you added an acceptance test for the functionality being added?
- [ ] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccConsul'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/... -v -run=TestAccConsul -timeout 120m
?   	github.com/hashicorp/terraform-provider-hcp/internal/clients	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/consul	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/input	(cached) [no tests to run]


=== RUN   TestAccConsulCluster
--- PASS: TestAccConsulCluster (553.00s)
=== RUN   TestAccConsulSnapshot
--- PASS: TestAccConsulSnapshot (591.61s)
PASS
ok  	github.com/hashicorp/terraform-provider-hcp/internal/provider	1144.969s
```

### CODE REVIEW

File: internal/provider/resource_consul_cluster.go
Comment (i0rek) [reply internal/provider/resource_consul_cluster.go:135]: I am honestly not sure what to do with this. There are two different things that can happen:

* false -> true: recreating the cluster will setup the peering
* true -> false: the already existing peering will be left untouched and nothing changes.

File: internal/provider/validators_test.go
Comment (i0rek) [reply internal/provider/validators_test.go]: done!

File: examples/resources/hcp_consul_cluster/federation.tf
Comment (i0rek) [reply examples/resources/hcp_consul_cluster/federation.tf:15]: Good point, I will change it!

File: internal/provider/resource_consul_cluster.go
Comment (i0rek) [reply internal/provider/resource_consul_cluster.go:132]: re 2) we should default to `false` as discussed with the platform-team.
re 3) personally I think we should merge this PR whenever it is ready. There is a LD flag for it that is being checked and I could make that always return true. Right now it is only enabled in dev env.

**Comment by i0rek** [comment]
I still need to generate the docs!

File: go.mod
Comment (bcmdarroch) [inline go.mod:13]: just cut [v0.10.0](https://github.com/hashicorp/hcp-sdk-go/releases/tag/v0.10.0) with your update

File: internal/provider/resource_consul_cluster.go
Comment (bcmdarroch) [reply internal/provider/resource_consul_cluster.go:132]: re 2/4. If the API sets the default value, you should add the line `Computed: true` to indicate the value will be set during Create even if the user doesn't provide a value.

re 3. While I don't foresee this happening, it is always possible we might need to cut an emergency release of the provider to fix a bug. So if this merged before the HVN peering resource was done, and then got pulled into a bug fix, it could be awkward. So maybe worth holding off merging until HVN peering is in just to be safe.

File: internal/provider/resource_consul_cluster.go
Comment (bcmdarroch) [reply internal/provider/resource_consul_cluster.go:136]: I'm also a little wary of using ForceNew here. In particular for the `true -> false`, it would be nice if we didn't force the user to recreate the clusters just because they've decided to do their own networking setup. Might be worth pulling in someone from product/design to discuss the possible user flows 🤔

**Review by bcmdarroch — COMMENTED** [review]
Looks good so far! I'd echo all of @riddhi89's comments, and I'd like to get some more product/design input on how updates to auto_peering should be handled.

File: internal/provider/resource_consul_cluster.go
Comment (i0rek) [reply internal/provider/resource_consul_cluster.go:132]: @bcmdarroch Do I need to set `Computed: true` when terraform and the API uses the default `false`?

Comment (bcmdarroch) [reply internal/provider/resource_consul_cluster.go:132]: Yes, I think so, if only to be explicit that it's set by the API.

File: internal/provider/resource_consul_cluster.go
Comment (bcmdarroch) [inline internal/provider/resource_consul_cluster.go:135]: Oh I realize what you meant by your other question. You can remove `Default` because `Computed` is set. `Default` is the provider picking a default value, `Computed` is the API picking a default value. Generally it's best to just let the API set all defaults, so you don't need to worry about keeping the terraform defaults in sync with the backend.
```suggestion

```

File: internal/provider/resource_consul_cluster.go
Comment (bcmdarroch) [inline internal/provider/resource_consul_cluster.go:461]: Does the Consul create operation include the HVN peering creation workflows? So once we receive the operation saying the Consul creation is complete, we can assume all peerings have been created too?

If so, I think a comment would be useful to share that hidden context.

**Comment by i0rek** [comment]
I removed the default, and re-added `ForceNew`.

File: templates/resources/consul_cluster.md.tmpl
Comment (bcmdarroch) [inline templates/resources/consul_cluster.md.tmpl:12]: This could be merged with above:
```suggestion
-> **Note:** The `auto_hvn_to_hvn_peering` and `primary_link` attributes are related to federation, a feature that is currently in private beta.
```

File: examples/resources/hcp_consul_cluster/federation.tf
Comment (bcmdarroch) [inline examples/resources/hcp_consul_cluster/federation.tf:26]: Does `auto_hvn_to_hvn_peering` have to be set to true on all Consul clusters in a federation? Or just the secondary? Worth calling out in the guide.

File: docs/guides/consul-federation.md
Comment (bcmdarroch) [inline docs/guides/consul-federation.md:13]: I'm thinking we could expand this guide. So this example could have the heading `Federation with Auto HVN Peering`, and show example HCL using auto peering. And then we can later add another section with a manual HVN peering example.

**Comment by i0rek** [comment]
Running integration tests right now.

File: examples/resources/hcp_consul_cluster/federation.tf
Comment (bcmdarroch) [inline examples/resources/hcp_consul_cluster/federation.tf:1]: I'm not sure this example is getting rendered anywhere in the docs 🤔 I'd move this into `examples/guides/consul_cluster_federation/main.tf `

**Review by bcmdarroch — COMMENTED** [review]
This is close! I have a few more questions, and some doc suggestions. I don't have all the feature flag set up for testing this locally, so I'll leave that to you to verify :)

File: internal/provider/resource_consul_cluster.go
Comment (i0rek) [reply internal/provider/resource_consul_cluster.go:461]: Yeah, that's the case. Where would you like this comment to go? Right where you commented? Or in the user facing docs?

File: examples/resources/hcp_consul_cluster/federation.tf
Comment (i0rek) [reply examples/resources/hcp_consul_cluster/federation.tf:26]: Good point! Will add a note in the guide and maybe in the docs as well.

**Comment by bcmdarroch** [comment]
Ah I did see a panic on my run of the acceptance test:
```
panic: Invalid address to set: []string{"auto_hvn_to_hvn_peering"}

goroutine 915 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc000c42380, 0x1ee0956, 0x17, 0x1ce92a0, 0x278c6e0, 0x0, 0x0)
	$GOPATH/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.5.0/helper/schema/resource_data.go:230 +0x371
github.com/hashicorp/terraform-provider-hcp/internal/provider.setConsulClusterResourceData(0xc000c42380, 0xc000b99380, 0xc0007aab90, 0xc00094fdd0, 0xc000b94450)
	$GOPATH/src/github.com/hashicorp/terraform-provider-hcp/internal/provider/resource_consul_cluster.go:462 +0x85b
github.com/hashicorp/terraform-provider-hcp/internal/provider.dataSourceConsulClusterRead(0x20ebd98, 0xc000b98240, 0xc000c42380, 0x1cbd9a0, 0xc000d6e0b0, 0xc000667f40, 0xc00076f948, 0x100fff8)
	$GOPATH/src/github.com/hashicorp/terraform-provider-hcp/internal/provider/data_source_consul_cluster.go:171 +0x5d6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc000148000, 0x20ebd28, 0xc000d14280, 0xc000c42380, 0x1cbd9a0, 0xc000d6e0b0, 0x0, 0x0, 0x0)
```

File: internal/provider/resource_consul_cluster.go
Comment (i0rek) [reply internal/provider/resource_consul_cluster.go:461]: I put a note right where you commented, because I don't think that's relevant for the user.

File: docs/guides/consul-federation.md
Comment (riddhi89) [inline docs/guides/consul-federation.md:37]: Add `auto_hvn_to_hvn_peering = "true"` here.

File: examples/guides/consul_cluster_federation/main.tf
Comment (riddhi89) [inline examples/guides/consul_cluster_federation/main.tf:23]: Missing `auto_hvn_to_hvn_peering = "true"` here.

File: internal/provider/resource_consul_cluster.go
Comment (riddhi89) [inline internal/provider/resource_consul_cluster.go:305]:
```suggestion
	// Enabling auto peering will peer this cluster's HVN with every other HVN with members in this federation.
```

File: internal/provider/resource_consul_cluster.go
Comment (riddhi89) [inline internal/provider/resource_consul_cluster.go:136]: Maybe you are still working on the PR - but based on the above feedback, add `Computed: true`?

File: templates/resources/consul_cluster.md.tmpl
Comment (bcmdarroch) [inline templates/resources/consul_cluster.md.tmpl:10]:
```suggestion
-> **Note:** The `primary_link` and `auto_hvn_to_hvn_peering` attributes are related to federation, a feature that is currently in private beta.
```

File: templates/guides/consul-federation.md.tmpl
Comment (bcmdarroch) [inline templates/guides/consul-federation.md.tmpl:11]:
```suggestion
By providing `auto_hvn_to_hvn_peering` as a parameter on the secondary cluster, the HVNs are peered
```

**Review by bcmdarroch — APPROVED** [review]
Nice work! Just two last few nits but otherwise this looks good to go! And acceptance tests are passing now 👍 👍

**Comment by i0rek** [comment]
@bcmdarroch yeah the panic was caused by the missing changes related to the data-source.

**Comment by i0rek** [comment]
@bcmdarroch @riddhi89 thank you both for your reviews! I think it is done.

**Review by riddhi89 — APPROVED** [review]
Tested it out locally. LGTM :shipit:

**Pull request merged**
